### ISSUE TITLE
Relational analysis (and potentially others) unsound due to eval on pre-state in function calls

### ISSUE THREAD

**Issue by michael-schwarz** [goblint/analyzer#1009]
Consider the following program

~~~C
// SKIP PARAM: --set ana.activated[+] apron --set ana.base.privatization none --set ana.relation.privatization top
#include <goblint.h>

int g = 0;
int h = 0;
int *ptr = &g;

int change_ptr_and_return_5() {
    ptr = &h;
    return 5;
}

int main() {
    // ptr points to h here, but apron evaluates on pre-state of function call, updating g instead
    *ptr = change_ptr_and_return_5();

    int x = g + 1; // Currently this fails because of operation on bot
    assert(g == h); //FAIL
    assert(h == 5);
    assert(g == 0);
}
~~~

Here, the value of `ptr` is changed inside the function, which happens before the assignment. However, we currently use the pre-state of base to resolve the pointer information. This is unsound.

In the example, this leads to:
- our relational analysis wrongly claiming that `g=5` and `h=0` and
- base correctly claiming `h=5` and `g=0`. 

This inconsistent state can lead to unsound analysis results & crashes in places such as assignments where the analyses aim to benefit from each other and then find the value of a variable to be `\bot`.

**Comment by sim642** [comment]
This is rather bad and a general gap in our approach. We would need every function call to first assign the return value to a fresh local temporary and then assign that to the actual lval to fix it. That way the `combine` is done before an assignment to a complex lval, which might depend on the state itself. It must be fresh because a local variable lval might be escaped (which isn't known a priori and also depends on the state).

## Race analysis
Similar issues arise from our race analysis where the must lockset is wrong if the function changes it.

### Imprecise
In one direction, this makes us imprecise as we find a race that's not there by having a smaller lockset before the call:
```c
#include <pthread.h>

int myglobal;
pthread_mutex_t mutex1 = PTHREAD_MUTEX_INITIALIZER;

int foo() {
  pthread_mutex_lock(&mutex1);
  return myglobal+1; // NORACE
}

void *t_fun(void *arg) {
  myglobal = foo(); // NORACE
  pthread_mutex_unlock(&mutex1);
  return NULL;
}

int main(void) {
  pthread_t id;
  pthread_create(&id, NULL, t_fun, NULL);
  pthread_mutex_lock(&mutex1);
  myglobal=myglobal+1; // NORACE
  pthread_mutex_unlock(&mutex1);
  pthread_join (id, NULL);
  return 0;
}
```

### Unsound
In the other direction, this makes us unsound as we find no race that's actually there by having a larger lockset before the call:
```c
#include <pthread.h>

int myglobal;
pthread_mutex_t mutex1 = PTHREAD_MUTEX_INITIALIZER;

int foo() {
  int x = myglobal + 1; // NORACE
  pthread_mutex_unlock(&mutex1);
  return x;
}

void *t_fun(void *arg) {
  pthread_mutex_lock(&mutex1);
  myglobal=foo(); // RACE!
  return NULL;
}

int main(void) {
  pthread_t id;
  pthread_create(&id, NULL, t_fun, NULL);
  pthread_mutex_lock(&mutex1);
  myglobal=myglobal+1; // RACE!
  pthread_mutex_unlock(&mutex1);
  pthread_join (id, NULL);
  return 0;
}
```

**Comment by sim642** [comment]
My idea for a fix is to split `combine` into two transfer functions:
1. `combine_environment` — combines all the global state.
2. `combine_assign` — performs the actual assign of the return value. Then the lval for the assign can be evaluated in the combined environment with the right globals.
